### PR TITLE
Add support for SAMD21 / Minitronics v2.0

### PR DESCRIPTION
**Pull request by brupje**
### Description

This PR adds support for SAMD21 compatible devices, specifically the Minitronics v2.0 by Reprapworld. 
New PR for cleaner commit history and adding support for saving settings to internal flash. 

Working:

- 3D printing
- CNC milling
- LCD support (hardware SPI)
- SD support (hw/sw SPI)
- Saving settings to internal flash
- Watchdog

### Requirements

Tested on Minitronics v2.0 only

### Benefits

Supports SAMD21 compatible devices

### Configurations

[Minitronics.2.0.zip](https://github.com/MarlinFirmware/Marlin/files/9983581/Minitronics.2.0.zip)

### CODE REVIEW

File: Marlin/src/HAL/SAMD21/HAL.h [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+#define __MSERIAL(X) MSerial##X
+#define _MSERIAL(X) __MSERIAL(X)
+#define MSERIAL(X) _MSERIAL(INCREMENT(X))
+
+#if SERIAL_PORT == -1
+  #define MYSERIAL1 MSerialUSB
+#else
+  #error "SERIAL_PORT must be -1 (Native USB only)."
+#endif
```
Comment (thinkyhead) [inline Marlin/src/HAL/SAMD21/HAL.h:61]: These blocks were originally specific to the Adafruit Grand Central serial ports. Other SAMD25 boards were not covered. So, these likely don't apply to the Minitronics 2.0, unless it really does only have a USB port and no other serial ports exposed. These should be changed to allow for other serial ports if others are available.

File: Marlin/src/inc/Conditionals_LCD.h [hunk 2]
```diff
@@ -137,6 +137,7 @@
   #define DOGLCD
   #define IS_U8GLIB_ST7920 1
   #define IS_ULTIPANEL 1
+  #define ENCODER_PULSES_PER_STEP 2
```
Comment (thinkyhead) [inline Marlin/src/inc/Conditionals_LCD.h:140]: You had some `REPRAPWORLD_GRAPHICAL_LCD` pins earlier, but they are not included in the new PR. Did you intend to remove them?

Comment (brupje) [reply Marlin/src/inc/Conditionals_LCD.h:140]: I don' t think I had them in this place. Marlinui_dogm has been updated to allow hardware SPI, maybe that is what you mean?

File: Marlin/src/HAL/SAMD21/HAL.h (hunk 1 above)
Comment (brupje) [reply Marlin/src/HAL/SAMD21/HAL.h:61]: fair, just hadn't tested it. I added serial 1 and 2

**Comment by brupje** [comment]
Any update on the comments above? Anything that needs to happen before What needs to be done before we can merge this?

**Comment by quiret** [comment]
Looking forward to refactoring yet another HAL_SPI.cpp for the big update... 😉

https://github.com/MarlinFirmware/Marlin/pull/24911

I wish someone could compensate me the cost for getting this SAMD21 printer controller board that you enable.

**Comment by thinkyhead** [comment]
> Looking forward to refactoring yet another HAL_SPI.cpp for the big update

Merge this first, then? Pretty soon… after another review and maybe some tweaks.

**Comment by quiret** [comment]
@thinkhead I would not mind. If you do then I will adjust my PR to add compatibility. It won't be the greatest HAL SPI implementation since I don't have a test board but it will work, even allowing generic SPI TFT as usual.

**Pull request merged**